syntax = "proto3";

package ecliptix.proto.group;

message GroupMember {
  bytes member_id = 1;
  bytes account_id = 2;
  bytes app_instance_id = 3;
  bytes device_id = 4;

  enum DeviceType {
    MOBILE = 0;
    DESKTOP = 1;
  }

  DeviceType device_type = 5;
  bytes identity_public_key = 6;
  uint64 joined_timestamp_ms = 7;

  enum MemberRole {
    MEMBER = 0;
    ADMIN = 1;
    OWNER = 2;
  }

  MemberRole role = 8;
  bool is_active = 9;
}

message GroupMetadata {
  bytes group_id = 1;
  string group_name = 2;
  bytes creator_id = 3;
  uint64 created_timestamp_ms = 4;
  uint64 last_modified_timestamp_ms = 5;
  uint32 version = 6;

  enum GroupType {
    PRIVATE = 0;
    PUBLIC = 1;
    BROADCAST = 2;
  }

  GroupType type = 7;
  uint32 max_members = 8;
  optional string description = 9;
}

message GroupState {
  GroupMetadata metadata = 1;
  repeated GroupMember members = 2;
  bytes current_sender_key_generation = 3;
  uint64 message_counter = 4;
  bytes transcript_hash = 5;
  uint64 last_sender_key_rotation_ms = 6;

  message PendingOperation {
    enum OperationType {
      ADD_MEMBER = 0;
      REMOVE_MEMBER = 1;
      UPDATE_METADATA = 2;
      ROTATE_SENDER_KEY = 3;
    }

    OperationType type = 1;
    bytes operation_id = 2;
    bytes initiator_id = 3;
    uint64 timestamp_ms = 4;
    bytes operation_data = 5;
  }

  repeated PendingOperation pending_operations = 7;
}

message SenderKeyState {
  bytes group_id = 1;
  bytes sender_id = 2;
  bytes sender_key_seed = 3;
  uint64 chain_index = 4;
  uint64 generation = 5;
  uint64 created_timestamp_ms = 6;
  uint32 message_count = 7;
}

message SenderKeyDistribution {
  bytes group_id = 1;
  bytes sender_id = 2;
  bytes sender_key_seed = 3;
  uint64 chain_index_start = 4;
  uint64 generation = 5;
  uint64 created_timestamp_ms = 6;
  bytes signature = 7;
}

message GroupMessage {
  bytes group_id = 1;
  bytes sender_id = 2;
  bytes sender_device_id = 3;
  uint64 sender_chain_index = 4;
  uint64 sender_key_generation = 5;
  bytes encrypted_content = 6;
  bytes content_nonce = 7;
  bytes auth_tag = 8;
  bytes transcript_hash = 9;
  uint64 timestamp_ms = 10;
  uint64 message_sequence = 11;

  message Reference {
    bytes referenced_message_id = 1;
    bytes referenced_sender_id = 2;
    uint64 referenced_timestamp_ms = 3;
  }

  optional Reference reply_to = 12;
}

message DeviceSyncMessage {
  bytes sender_device_id = 1;
  bytes account_id = 2;
  GroupState group_state = 3;
  repeated SenderKeyState sender_keys = 4;
  repeated bytes recent_message_hashes = 5;
  uint64 sync_sequence = 6;
  uint64 timestamp_ms = 7;
}

message GroupInvite {
  bytes invite_id = 1;
  bytes group_id = 2;
  bytes inviter_id = 3;
  bytes invitee_account_id = 4;
  GroupMetadata group_metadata = 5;
  repeated bytes member_ids = 6;
  SenderKeyDistribution initial_sender_key = 7;
  uint64 expires_at_ms = 8;
  bytes signature = 9;
}

message MembershipChange {
  enum ChangeType {
    ADDED = 0;
    REMOVED = 1;
    LEFT = 2;
    ROLE_CHANGED = 3;
  }

  ChangeType change_type = 1;
  bytes group_id = 2;
  bytes member_id = 3;
  bytes initiator_id = 4;
  uint64 timestamp_ms = 5;
  optional GroupMember.MemberRole new_role = 6;
  bytes new_sender_key_generation = 7;
}
